app/src/lib/stores/git-store.ts

```diff
@@ -1114,7 +1114,7 @@ export class GitStore extends BaseStore {
   }
 
   /** Merge the named branch into the current branch. */
-  public merge(branch: string): Promise<void> {
+  public merge(branch: string): Promise<void | true> {
```

Comment:
I'm wrapping my head around these internals changed but I have some suggestions for this code path:

 - omit `true` from the inner `Promise<T>` return type on the Git operation (if the action fails, the promise is rejected and that needs to be handled):

```diff
diff --git a/app/src/lib/git/merge.ts b/app/src/lib/git/merge.ts
index bccd9ee45..2e78d44b4 100644
--- a/app/src/lib/git/merge.ts
+++ b/app/src/lib/git/merge.ts
@@ -9,9 +9,8 @@ import { spawnAndComplete } from './spawn'
 export async function merge(
   repository: Repository,
   branch: string
-): Promise<void | true> {
+): Promise<void> {
   await git(['merge', branch], repository.path, 'merge')
-  return true
 }

 /**
```

 - change `GitStore` return a `boolean` to indicate whether the operation succeeded or not (internal error handling takes care of the rest):

```diff
diff --git a/app/src/lib/stores/git-store.ts b/app/src/lib/stores/git-store.ts
index 598416e30..cec479f93 100644
--- a/app/src/lib/stores/git-store.ts
+++ b/app/src/lib/stores/git-store.ts
@@ -1114,10 +1114,15 @@ export class GitStore extends BaseStore {
   }
 
   /** Merge the named branch into the current branch. */
-  public merge(branch: string): Promise<void | true> {
-    return this.performFailableOperation(() => merge(this.repository, branch), {
-      command: 'merge',
-    })
+  public async merge(branch: string): Promise<boolean> {
+    const result = await this.performFailableOperation(
+      () => merge(this.repository, branch),
+      {
+        command: 'merge',
+      }
+    )
+
+    return result !== undefined
   }
 
   /** Changes the URL for the remote that matches the given name  */
```
 
 - tidy up the `AppStore` checks (only show the dialog when you have a valid tip):

```diff
diff --git a/app/src/lib/stores/app-store.ts b/app/src/lib/stores/app-store.ts
index ffd92c175..b39108954 100644
--- a/app/src/lib/stores/app-store.ts
+++ b/app/src/lib/stores/app-store.ts
@@ -3079,13 +3079,12 @@ export class AppStore extends TypedBaseStore<IAppState> {
       }
     }
 
-    if ((await gitStore.merge(branch)) === true) {
-      const currentBranchName =
-        gitStore.tip.kind === TipState.Valid
-          ? gitStore.tip.branch.name
-          : 'unknown branch'
+    const mergeSuccessful = await gitStore.merge(branch)
+    const { tip } = gitStore
+
+    if (mergeSuccessful && tip.kind === TipState.Valid) {
       this._setSuccessfulMergeBannerState({
-        currentBranch: currentBranchName,
+        currentBranch: tip.branch.name,
         theirBranch: branch,
       })
     }
```

Reply:
```diff
+  public async merge(branch: string): Promise<boolean> {
+    const result = await this.performFailableOperation(
+      () => merge(this.repository, branch),
+      {
+        command: 'merge',
+      }
+    )
+
+    return result !== undefined
```

@shiftkey so i tried something like this first, but the part i got stuck on was differentiating a `void` return and an `undefined` return. i think the same issue is here in this implementation?

> tidy up the AppStore checks (only show the dialog when you have a valid tip):

👍 will make this change

Reply:
Possibly? I'll think about this some more and see if I can spot a non-ugly way out of this...

Reply:
@shiftkey yep that sounds about right